### PR TITLE
[Import Decl] Refactoring ObjCMethod importing

### PR DESCRIPTION
**Pull request by milseman**
[Import Decl] Don’t import as init using omit needless words

Previously, for an Objective-C class method declaration that could be
imported as init, we could make 4 decls, which included a class method decl using omit-needless-words. But, this was never a valid name anyways, nor is it easy for users to divine the magic omit-needless-words transformations that are done. So, we drop and treat the Swift 2 non-init name as a  "raw" name, which is the name as it appears from ObjC. This will align nicely with future ImportName refactoring, where ImportDecl becomes parameterized based on language version, or otherwise is passed the name ahead of time.

### CODE REVIEW

File: lib/ClangImporter/ImporterImpl.h [hunk 1]
```diff
@@ -440,19 +445,23 @@ class LLVM_LIBRARY_VISIBILITY ClangImporter::Implementation
...
 
-  /// A mapping from imported declarations to their "alternate" declarations,
-  /// for cases where a single Clang declaration is imported to two
-  /// different Swift declarations.
-  llvm::DenseMap<Decl *, ValueDecl *> AlternateDecls;
-
   /// Retrieve the alternative declaration for the given imported
   /// Swift declaration.
-  ValueDecl *getAlternateDecl(Decl *decl) {
+  TinyPtrVector<ValueDecl *> getAlternateDecls(Decl *decl) {
```
Comment (jrose-apple) [inline lib/ClangImporter/ImporterImpl.h]: Suggestion: ArrayRef.

File: test/IDE/print_omit_needless_words.swift [hunk 2]
```diff
@@ -53,8 +53,8 @@
 // Note: Don't drop the name of the first parameter in an initializer entirely.
 // CHECK-FOUNDATION: init(array: [Any])
 
-// Note: struct name matching; don't drop "With".
-// CHECK-FOUNDATION: class func withRange(_: NSRange) -> NSValue
+// Note: Init formed, no withRange member
+// CHECK-FOUNDATION:  /*not inherited*/ init!(range: NSRange)
```
Comment (jrose-apple) [inline test/IDE/print_omit_needless_words.swift]: I'm pretty sure we still want these test cases in some form…

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
My branch was out of date, rebasing and fixing now

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci thank you

**Comment by milseman** [comment]
Failure is probably un-related:
```
19:12:52 2  swiftc                   0x000000010cdcb9ea swift::DeclContext::lookupQualified(swift::Type, swift::DeclName, swift::NLOptions, swift::LazyResolver*, llvm::SmallVectorImpl<swift::ValueDecl*>&) const + 58
```

@slavapestov do you know why that is failing like that?

**Comment by milseman** [comment]
Awaiting https://github.com/apple/swift/pull/6020, then testing again

**Comment by gparker42** [comment]
It's a bug in #5990. We're testing a fix in #6020.

**Comment by gparker42** [comment]
@swift-ci please test OS X platform

**Comment by milseman** [comment]
@swift-ci please test

**Pull request merged**